src/lambda/commands/addSamDebugConfiguration.ts

```diff
@@ -0,0 +1,37 @@
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as vscode from 'vscode'
+import { LaunchConfiguration } from '../../shared/debug/launchConfiguration'
+import { createDirectInvokeSamDebugConfigurationFromTemplate } from '../../shared/sam/debugger/awsSamDebugger'
+
+export interface AddSamDebugConfigurationInput {
+    samTemplateResourceName: string
+    samTemplateUri: vscode.Uri
+}
+
+/**
+ * Adds a new debug configuration for the given sam function resource and template.
+ */
+export async function addSamDebugConfiguration({
+    samTemplateResourceName: resourceName,
+    samTemplateUri: templateUri
+}: AddSamDebugConfigurationInput): Promise<void> {
```

Comment:
1.

Yes, I agree in most cases. However for Commands, you lose all static typing information. They take in an array of `any`. So the best way to create a contract for a Command seems to be to create an input container. I actually saw this being done elsewhere in the codebase.

side-note: it's generally considered best practice to create a container if you have a few parameters of the same type as well. In a language without named parameters like ts/js, it's extremely easy to create bugs by getting the order of the parameters wrong. Creating a type prevents those bugs and creates self-documenting code at the call site

e.g. reading and writing `foo(true, false)` vs. reading and writing `foo({doThing: true, doOtherThing: false})`

---

2.

New to destructuring and not familiar with best practices here. But it *seems* like it makes sense to destructure whenever you continuously end up repeating yourself by referring to the parent. 

This ties into (1). It lets you get the best of both worlds by having a container object (enforces a contract, reduces bugs, better readability at call site) while also reducing verbosity in the body of the function (better readability within body).

Reply:
> They take in an array of `any`. So the best way to create a contract for a Command seems to be to create an input container.

Adding and _naming_ interfaces is a lot of noise just to avoid ordering bugs. Even one insists on this cost, the interface can be inline, it doesn't need a name.

Reply:
I think I understand the justification in this case: that having a common exportable interface forces the types to match on both sides of the VS Code command-barrier.

Not sure if I'm a fan of the destructuring if you're already using the interface but that's not necessarily a battle I care to fight. Maybe just shorten the interface names to match the variable names so you can get away with something like this (in theory, the `samTemplate` context should be provided by the interface's name):

```
export async function addSamDebugConfiguration({
    resourceName, templateUri
}: AddSamDebugConfigurationInput): Promise<void> {
```

Reply:
This is a heavy cost to avoid a problem that would be caught by a basic acceptance test.

Interface and class names occupy the global namespace of the project. When 80% of them are useless noise, it makes the 20% of actual meaning difficult to find and understand.

Reply:
I would disagree that interfaces are a high cost or noise, especially in typescript. It allows for much better self-documenting code and, in this case, sanity type checks at the call site. I'm not a fan of the inlined interfaces but perhaps that will change. I find them hard to read at the declaration site and when writing code. I far prefer a well-documented interface that conveys the intent of the type.

Considering the heavy use of interfaces in the vs code apis, it seems idiomatic. But I'm new to typescript so I haven't grasped best practices yet. I would however be very surprised if people argued against explicit interfaces.

I've also never had a problem with auto-suggest pollution except when multiple dependencies are using the same type name like `List`.